Feat/explicit conversion operator

theodelrieu opened the pull request:
Here it is!

Thanks to @gracicot, who found a good solution to #958, I'm finally attempting to solve it.

This PR introduces a CMake option: `JSON_ExplicitConversions`, defaulted to `OFF`.

To help users move away from the implicit conversion operator, a warning will be issued each time it is used. This warning can be silenced by users defining `JSON_USE_IMPLICIT_CONVERSIONS`.

On the other hand, defining `JSON_USE_EXPLICIT_CONVERSIONS` will very likely break code, but will ease the transition to the next major version, where only explicit conversion operator will be kept.

Meanwhile, I've removed every implicit conversion from the library's code, including tests.
It might be a good idea to add a short `ifdef`ed test case, which tests implicit conversions (to avoid duplicating a lot of test code).

With explicit conversions, we can also remove the constraints that are present today, e.g. forbidding conversion from `char`, `std::initializer_list<char>`. Note that this has not been done in this PR.

I think this change is more suited to be released in a minor version, so we might release 3.6.2 first?

## Pull request checklist

Read the [Contribution Guidelines](https://github.com/nlohmann/json/blob/develop/.github/CONTRIBUTING.md) for detailed information.

- [x]  Changes are described in the pull request, or an [existing issue is referenced](https://github.com/nlohmann/json/issues).
- [x]  The test suite [compiles and runs](https://github.com/nlohmann/json/blob/develop/README.md#execute-unit-tests) without error.
- [x]  [Code coverage](https://coveralls.io/github/nlohmann/json) is 100%. Test cases can be added by editing the [test suite](https://github.com/nlohmann/json/tree/develop/test/src).
- [x]  The source code is amalgamated; that is, after making changes to the sources in the `include/nlohmann` directory, run `make amalgamate` to create the single-header file `single_include/nlohmann/json.hpp`. The whole process is described [here](https://github.com/nlohmann/json/blob/develop/.github/CONTRIBUTING.md#files-to-change).

theodelrieu commented:
Where is the LGTM config file?

gracicot commented on include/nlohmann/detail/macro_scope.hpp:
```diff
@@ -153,3 +153,14 @@
+#ifndef JSON_USE_EXPLICIT_CONVERSIONS
+    #define JSON_EXPLICIT
+    #ifndef JSON_USE_IMPLICIT_CONVERSIONS
+        #warning Implicit conversions are being deprecated and will be \
```
I realized that the `#warning` statement is not standard, and not supported on msvc. There is some solution like `#pragma message()` but it's not a real warning.

I think we could use `[[deprecated("...")]]` or the msvc non standard `__declspec(deprecated("..."))` for earlier versions in that case.

theodelrieu replied:
I've added JSON_DEPRECATED_MSG, and tried building the tests by removing the `Wno-deprecated` warnings, it works \o/

gracicot replied:
Thank you, the patch is great now. I would just wait for @nlohmann to do a review too and merge it!

nlohmann replied:
I like the idea of having a switch for the conversions, but I am still not convinced that implicit conversions should be deprecated.

theodelrieu replied:
The main reason is because it is broken today. The pile of issues related to #958 is a good indicator of that. Not to mention the **why** is quite hard to grasp.

The required changes are quite minimal as well:

```cpp
std::string s;
s = j;
// becomes
std::string s;
j.get_to(s);

std::string s = j;
// becomes
std::string s(j);
```

gracicot replied:
After thinking about it I think @nlohmann is right on not deprecate *right now* (I still think it should eventually). The compile switch must be there for a while before any deprecation, and the documentation should also rely on explicit conversion and it should recommend enabling the switch.

There is non-trivial cases that must be refactored when changing from implicit to explicit. Take this for example:
```c++
struct special_type {
    int a, b;
    std::string c, d;
};

special_type{value1, value2, value3, value4}; // works with implicit only
```

I really don't think a lot of users are relying on this, but it's still an annoyance. If we can recommend using the compiler switch for a while, warn users that the implicit operator will be deprecated in the future then the transition will be smooth.

The only way to recreate the behaviour above is to create a thin wrapper that allows casts:

```c++
struct implicit {
    json::value value;
    template<typename T, typename = decltype(std::declval<json::value>().get<T>())>
    operator T () {
        return value.get<T>();
    }
};

special_type{implicit{value1}, implicit{value2}, implicit{value3}, implicit{value4}};
```

I would like to see the implicit operator be deprecated too but I also agree that even though the change in the syntax is minimal, its use is widespread so it must be done very gradually.

nlohmann commented:
> Where is the LGTM config file?

There is no config file. Here is a link to the log: https://lgtm.com/projects/g/nlohmann/json/logs/rev/pr-e7a2ac9be19e57d05404374d62bf6a8ab3794bfe/lang:cpp/stage:Build%20merge_d94187b754143b1064d1f7bcd3000e3db0419555. Strange that the test fails.

theodelrieu commented:
But where do you define how to build the library?

theodelrieu commented:
Ok, the LGTM build uses default CMake arguments, therefore it builds with the implicit conversions.
All the other builds use explicit conversions.

@nlohmann do you wish to do like with `JSON_MultipleHeaders`? i.e. define a variable in the CI files.

theodelrieu commented:
It would be nice to merge this PR and discuss what is the long-term plan after.

theodelrieu commented:
@nlohmann Shall we merge this PR now that 3.7.0 has been released?

gracicot commented:
Since tackling the conversion operator is being considered for version 4.0.0, I'd vote to add the deprecation back when `JSON_ExplicitConversions` is off.

If we want to use explicit conversion on the next major version, deprecation and the switch is the first step to transition the community.

gregmarr commented:
@theodelrieu All of the whitespace and line wrapping changes in the first commit makes it really hard to see what's changed, even with the "ignore whitespace" option.  They also broke the formatting of some of the tables in the comments.  https://github.com/nlohmann/json/pull/1559/files#diff-641eb804df1899b4e8277af5a52e6b7dR335

theodelrieu commented:
@gregmarr damnit, I'll run amalgamate in a first commit

nlohmann commented on CMakeLists.txt:
```diff
@@ -24,6 +24,7 @@ endif ()
 option(JSON_BuildTests "Build the unit tests when BUILD_TESTING is enabled." ON)
 option(JSON_Install "Install CMake targets during install step." ON)
 option(JSON_MultipleHeaders "Use non-amalgamated version of the library." OFF)
+option(JSON_ExplicitConversions "Enable explicit conversion operator" OFF)
```
I'm not sure about the wording. Isn't the flag more about *disabling* implicit conversions? At least this would be the difference between the current code and when the flag is set.

gregmarr replied:
The technical side is making the conversion operator explicit.  If you're looking at the option from the POV of how it changes behavior, then yes it is disabling implicit conversions, but if you are looking at the effect the option has on the code, then making the conversion operator explicit makes sense.  (This assumes that this is all that this option does.  If it's doing more than that, then it may make sense to word it the other way.)

gregmarr replied:
The next question is what happens to this option in the next major version where we can make this the default?  Does it become "Enable implicit conversion operator"?

nlohmann replied:
Hm. To me, an explicit conversion seems less a feature than an implicit one. So I would rather expect a switch controlling whether or not I would like to have implicit conversions. So for the time being,

```cmake
option(JSON_ImplicitConversions "Enable implicit conversion operators" ON)
```

theodelrieu replied:
@nlohmann The current behavior is broken, I wouldn't reason in terms of which one is more a feature, but rather as a new behavior

gregmarr replied:
Ah, I wasn't thinking of the possibility of having the "use CMake" option being different than the "just include the header" option.  That would make it so that we don't have to change the name of the CMake option in the next major version.

nlohmann replied:
I'm still struggling with the wording. To me, the switch "enable explicit conversion" sounds like an additional feature. Instead, it breaks existing code, because the assumed implicit conversions do not work any longer. If the code already only used explicit conversions, then flipping the switch does not change anything.

nlohmann commented on .travis.yml:
```diff
@@ -318,7 +318,7 @@ script:
   # compile and execute unit tests
   - mkdir -p build && cd build
-  - cmake .. ${CMAKE_OPTIONS} -DJSON_MultipleHeaders=${MULTIPLE_HEADERS} -DJSON_BuildTests=On -GNinja && cmake --build . --config Release
+  - cmake .. ${CMAKE_OPTIONS} -DJSON_ExplicitConversions=ON -DJSON_MultipleHeaders=${MULTIPLE_HEADERS} -DJSON_BuildTests=ON -GNinja && cmake --build . --config Release
```
Why do you set the flag for all Travis tests? Wouldn't it be better to have it set for one test - just like what we do with disabled exceptions?

nlohmann commented on Makefile:
```diff
@@ -71,7 +71,7 @@ check:
 	rm -fr build_coverage
 	mkdir build_coverage
-	cd build_coverage ; cmake .. -GNinja -DCMAKE_BUILD_TYPE=Debug -DJSON_Coverage=ON -DJSON_MultipleHeaders=ON
+	cd build_coverage ; cmake .. -GNinja -DCMAKE_BUILD_TYPE=Debug -DJSON_Coverage=ON -DJSON_MultipleHeaders=ON -DJSON_ExplicitConversions=ON
```
Same question - why to set this flag for coverage?

nlohmann commented on appveyor.yml:
```diff
@@ -112,7 +112,7 @@ before_build:
   - ps: if ($env:name -Eq "with_win_header") { $header_path = "single_include\nlohmann\json.hpp" }
   - ps: if ($env:name -Eq "with_win_header") { "#include <Windows.h>`n" + (Get-Content $header_path | Out-String) | Set-Content $header_path }
-  - if "%GENERATOR%"=="Ninja" (cmake . -G "%GENERATOR%" -DCMAKE_BUILD_TYPE="%configuration%" -DCMAKE_CXX_FLAGS="%CXX_FLAGS%" -DCMAKE_EXE_LINKER_FLAGS="%LINKER_FLAGS%" -DCMAKE_IGNORE_PATH="C:/Program Files/Git/usr/bin" -DJSON_BuildTests=On) else (cmake . -G "%GENERATOR%" -A "%GENERATOR_PLATFORM%" -DCMAKE_CXX_FLAGS="%CXX_FLAGS%" -DCMAKE_EXE_LINKER_FLAGS="%LINKER_FLAGS%" -DCMAKE_IGNORE_PATH="C:/Prog ...
+  - if "%GENERATOR%"=="Ninja" (cmake . -G "%GENERATOR%" -DCMAKE_BUILD_TYPE="%configuration%" -DCMAKE_CXX_FLAGS="%CXX_FLAGS%" -DCMAKE_EXE_LINKER_FLAGS="%LINKER_FLAGS%" -DCMAKE_IGNORE_PATH="C:/Program Files/Git/usr/bin" -DJSON_BuildTests=On -DJSON_ExplicitConversions=On) else (cmake . -G "%GENERATOR%" -A "%GENERATOR_PLATFORM%" -DCMAKE_CXX_FLAGS="%CXX_FLAGS%" -DCMAKE_EXE_LINKER_FLAGS="%LINKER_FLAGS% ...
```
Same question as above - why setting the flag as default for the CIs?

nlohmann commented on include/nlohmann/detail/conversions/from_json.hpp:
```diff
@@ -144,16 +143,20 @@ void from_json(const BasicJsonType& j, std::forward_list<T, Allocator>& l)
     }
     l.resize(j.size());
-    std::copy(j.begin(), j.end(), std::begin(l));
+    std::transform(
```
What's the benefit of using transform rather than copy?

gregmarr replied:
copy requires implicit conversions

nlohmann commented on include/nlohmann/json.hpp:
```diff
@@ -319,12 +332,16 @@ class basic_json
-    `url`       | The URL of the project as string.
-    `version`   | The version of the library. It is an object with the following keys: `major`, `minor`, and `patch` as defined by [Semantic Versioning](http://semver.org), and `string` (the version string).
+    `compiler`  | Information on the used compiler. It is an object with the
+    following keys: `c++` (the used C++ standard), `family` (the compiler family;
```
This looks broken.

nlohmann commented on include/nlohmann/json.hpp:
```diff
@@ -1049,7 +1069,9 @@ class basic_json
                     {
-                        JSON_THROW(other_error::create(500, "961c151d2e87f2686a955a9be24d316f1362bf21 3.8.0")); // LCOV_EXCL_LINE
+                        JSON_THROW(other_error::create(500,
+                                                       "961c151d2e87f2686a955a9be24d316f136"
```
Please revert this change - the hash should be in one literal.

nlohmann commented on include/nlohmann/json.hpp:
```diff
@@ -1262,14 +1286,21 @@ class basic_json
-    parse_event_t::value | the parser finished reading a JSON value | depth of the value | the parsed JSON value
-
-    @image html callback_events.png "Example when certain parse events are triggered"
+    parse_event_t::object_start | the parser read `{` and started to process a
```
Please revert the whitespace changes in comments.

nlohmann commented on include/nlohmann/json.hpp:
```diff
@@ -1336,8 +1367,7 @@ class basic_json
     */
-    basic_json(const value_t v)
-        : m_type(v), m_value(v)
+    basic_json(const value_t v) : m_type(v), m_value(v)
```
Style: please revert this change.

nlohmann commented on include/nlohmann/json.hpp:
```diff
@@ -1360,8 +1390,7 @@ class basic_json
     */
-    basic_json(std::nullptr_t = nullptr) noexcept
-        : basic_json(value_t::null)
+    basic_json(std::nullptr_t = nullptr) noexcept : basic_json(value_t::null)
```
Style: please revert this change.

nlohmann commented on include/nlohmann/json.hpp:
```diff
@@ -3215,18 +3306,25 @@ class basic_json
+            && detail::is_detected<detail::get_template_function,
+                                   const basic_json_t&, ValueType>::value,
+            int >::type = 0 >
+    JSON_EXPLICIT operator ValueType() const
```
As this is the only place where `JSON_EXPLICIT` is actually used, please revert all the whitespace/formatting changes to have a better overview what actually changed.

nlohmann commented on include/nlohmann/json.hpp:
```diff
@@ -3793,20 +3898,23 @@ class basic_json
             if (it != end())
             {
-                return *it;
+                return it->template get<ValueType>();
```
Why this change?

gregmarr replied:
This code was using the implicit conversion operator.

nlohmann commented on test/src/unit-udt.cpp:
```diff
@@ -311,30 +311,6 @@ TEST_CASE("basic usage" * doctest::test_suite("udt"))
             person_json["name"].get_to(name).m_val = "new name";
             CHECK(name.m_val == "new name");
         }
-
```
I do not think this test should be removed. Can't we detect the flag?

theodelrieu replied:
It's been a while since I've worked on this PR, I'll try to make both cohabit. That's also why I've put `JSON_ExplicitConversions=ON` everywhere.

theodelrieu replied:
One annoying thing is that code using implicit conversions will not build when explicit conversions are enabled, I'd like to find a way to avoid sprinkling `ifdef` everywhere in the tests.

gregmarr commented on include/nlohmann/detail/conversions/from_json.hpp:
```diff
@@ -144,16 +143,20 @@ void from_json(const BasicJsonType& j, std::forward_list<T, Allocator>& l)
     l.resize(j.size());
-    std::copy(j.begin(), j.end(), std::begin(l));
+    std::transform(
+        j.begin(), j.end(), std::begin(l),
```
Can we use `std::begin` and `std::end` on `j` as well for consistency?

theodelrieu replied:
I don't see the need to do that, since `j` is of type `json`, thus we know its member functions (c.f. `is_array` used a few lines above).

Also, I think this change is outside the scope of this PR.

gregmarr replied:
It just seems jarring to mix member and non-member in the same function call.  If we weren't already changing this line, then I wouldn't bring it up.  In any case, not something for which I would hold up the PR.

theodelrieu commented:
Should be good (bis)

theodelrieu commented:
@nlohmann @gregmarr Some changes in the tests are necessary to make the code compile/work on both implicit/explicit conversions builds.

Containers having a `std::initializer_list` cannot be constructed with braces (i.e. `{json}`), because it will fail in implicit conversion builds (but will work in explicit conversion builds).

gregmarr commented on include/nlohmann/json.hpp:
```diff
@@ -3793,7 +3794,7 @@ class basic_json
             if (it != end())
             {
-                return *it;
+                return static_cast<ValueType>(*it);
```
Curious if there is a technical reason that we use `static_cast<ValueType>` here but `get<ValueType>` in other places.

theodelrieu replied:
good catch

gregmarr commented on include/nlohmann/json.hpp:
```diff
@@ -8333,8 +8334,8 @@ class basic_json
             // collect mandatory members
-            const std::string op = get_value("op", "op", true);
-            const std::string path = get_value(op, "path", true);
+            const std::string op(get_value("op", "op", true));
```
Is there a technical reason for this change (3 lines)?

theodelrieu replied:
The previous makes use of implicit conversions, which does not build when using the `explicit` operator.

gregmarr replied:
Ah, fun with different initialization syntaxes.

gregmarr commented on test/src/unit-cbor.cpp:
```diff
@@ -1011,21 +1011,21 @@ TEST_CASE("CBOR")
                     {
                         json j = json::from_cbor(std::vector<uint8_t>({0xf9, 0x00, 0x00}));
-                        json::number_float_t d = j;
+                        json::number_float_t d{j};
```
Is there a technical reason for this change, and if so, is this something that users of the library will have to do as well if they enable explicit conversions?

gregmarr replied:
I assume this is the same answer as the other one.  In that case, we should probably add something to the docs/README about the error that can happen and how to resolve it.

theodelrieu replied:
The previous line simply does not compile when using explicit conversions. There are several ways to fix it, the easiest is the one I wrote, but you can be really explicit and use `get`, or even `static_cast`.

theodelrieu commented:
Ok, those macOS fails are due to libcxx 5.0.2 `std::string` hidden constructor, taking a `__self_view` (i.e. `std::string_view`, even when not using C++17).

Thus testing if a json object is explicitly convertible to a `std::string` will fail to compile. I removed the `is_explicitly_convertible_to` trait, and replaced it with a `is_getable` trait.

This trait checks that `j.get<T>()` does not yield a compiler error. It should be used as a SFINAE condition wherever `get` is used (instead of `std::is_convertible`, or `has_from_json`).

theodelrieu commented:
@nlohmann In the last run, the single fail is due to a GitHub outage, can you rerun the failed job? Or do I have to trigger a build?

nlohmann commented:
Done.

theodelrieu commented:
@nlohmann Seems good finally :pray:

nlohmann commented on test/src/unit-conversions.cpp:
```diff
@@ -734,235 +542,430 @@ TEST_CASE("value conversion")
         SECTION("exception in case of a non-number type")
         {
-            CHECK_THROWS_AS(json(json::value_t::null).get<json::number_integer_t>(),
+            CHECK_THROWS_AS(json(json::value_t::null).get<json::boolean_t>(),
```
Why these changes to from `number_integer_t` to `boolean_t`?

theodelrieu replied:
Oops, must be a leftover from last year

nlohmann commented on include/nlohmann/detail/meta/type_traits.hpp:
```diff
@@ -94,6 +94,12 @@ using get_template_function = decltype(std::declval<T>().template get<U>());
 template<typename BasicJsonType, typename T, typename = void>
 struct has_from_json : std::false_type {};
 
+template <typename BasicJsonType, typename T>
```
Could you please add the rationale why this function is needed rather than `std::is_constructible`?

theodelrieu replied:
I've added the rationale [here](https://github.com/nlohmann/json/pull/1559#issuecomment-661111824)

theodelrieu replied:
Do you want me to add it in the code?

nlohmann replied:
Yes, please. Such things are so easy forgotten...

nlohmann commented on test/src/unit-conversions.cpp:
```diff
@@ -189,48 +189,6 @@ TEST_CASE("value conversion")
         }
     }
 
-    SECTION("get an object (implicit)")
```
Maybe we should add more `#ifdef` blocks rather than reordering the file?

theodelrieu replied:
What is the benefit?

nlohmann replied:
The benefit is to maintain the order of the tests.

theodelrieu commented:
@nlohmann About your last comment: I can swap the CMake option definition

theodelrieu commented:
@nlohmann GitHub was unreachable again, sigh. Can you restart the job please? :pray: (https://travis-ci.org/github/nlohmann/json/jobs/710669524#L5359)

theodelrieu commented:
@nlohmann It seems appveyor is stuck as well 🤔

nlohmann commented:
No, it's just slow...

aokellermann commented on CMakeLists.txt:
```diff
@@ -24,6 +24,7 @@ endif ()
 option(JSON_BuildTests "Build the unit tests when BUILD_TESTING is enabled." ON)
 option(JSON_Install "Install CMake targets during install step." ON)
 option(JSON_MultipleHeaders "Use non-amalgamated version of the library." OFF)
+option(JSON_ImplicitConversions "Enable implicit conversions" ON)
```
Maybe put period at end of description for consistency's sake.

gracicot commented:
I wonder if this can be removed to re-enable conversions to `std::string_view` once we switch to explicit conversions:
```
    && !std::is_same<ValueType, typename std::string_view>::value
```
I think it was disabled because of a problem with string views and visual studio, but I'm not sure. If someone can find the issue/pull request, that would be awesome.

nlohmann commented:
@gracicot Was this changed in this PR?

gracicot commented:
> @gracicot Was this changed in this PR?

No.

nlohmann commented:
> No. The pull request mentioned that this hasn't been done for`std::initializer_list` but I was wondering if `std::string_view` was worth to do. Right now `json_value.get<std: string_view>()` works, but not the conversion and `is_gettable` returns false for it.

I see. Let's handle this in a follow-up issue. This PR has been open for too long...

nlohmann commented:
Thank you all so much!

nlohmann commented:
* * *

### 🔖 Release item

This issue/PR will be part of the next release of the library. This template helps preparing the release notes.

#### Type

- [x] ✨ New Feature
- [ ] 🐛 Bug Fix
- [ ] ⚡️ Improvement
- [ ] 🔨 Further Change
- [ ] 🔥 Deprecated function

#### Description

- Add option to switch off implicit conversions. #958 #1559

* * *

theodelrieu commented:
@gracicot I think all these type of explicit SFINAE disabling can be removed when using explicit conversions. IIRC there is one for `char` as well somewhere.

gracicot commented:
Thank you so much! 🎉

tlemo commented:
Thanks @theodelrieu for driving this!
